run-tests.sh

```diff
@@ -61,19 +62,69 @@ compile_pnut() {
   fi
 }
 
+shell_version() {
+  case "$1" in
+    bash) bash -c 'echo $BASH_VERSION' ;;
+    ksh)  ksh  -c 'echo $KSH_VERSION' ;;
+    mksh) mksh -c 'echo $KSH_VERSION' ;;
+    yash) yash --version | head -n 1 ;;
+    zsh)  zsh  --version ;;
+    # dash doesn't support --version or DASH_VERSION and we'd have to query the package manager
+    dash) echo "dash" ;;
+    *)    echo "Unknown shell: $1" ;;
+  esac
+}
+
 # Some tests specify command line arguments in the source file
 # This function extracts the arguments from the source file
 # To specify arguments, add a comment in the source file like this:
 # // args: arg1 arg2 arg3
 test_args() {
-  echo `sed -n -e "/\/\/ args:/p" "$1" | sed -e "s/^\/\/ args://" |  tr '\n' ',' | sed -e 's/,$//'`
+  echo `sed -n -e "/\/\/ args: /p" "$1" | sed -e "s/^\/\/ args: //" |  tr '\n' ',' | sed -e 's/,$//'`
+}
+
+# Some shells don't support certain features which mean some tests will fail.
+# While we often can work around the bugs and non-standard behavior of certain
+# shells, it can be easier to just disable the tests, especially if the test is
+# not relevant to the bootstrap process.
+# // expect_failure_for: bash-2*
+# // expect_failure_for: yash
+test_expect_falure_for_shells() {
+  echo `sed -n -e "/\/\/ expect_failure_for: /p" "$1" | sed -e "s/^\/\/ expect_failure_for: //"`
+}
+
+# Some tests take a long time to run, so we set a timeout to prevent infinite
+# loops However, we don't want to set a high timeout for all tests, so we have
+# an option to set a specific timeout.
+test_timeout() {
+  echo `sed -n -e "/\/\/ timeout: /p" "$1" | sed -e "s/^\/\/ timeout: //"`
+}
+
+test_expect_failure_for_shell() { # file: $1
+  failing_shells=$(test_expect_falure_for_shells "$1")
+  for failing_shell in $failing_shells; do
+    failing_shell_name=$(echo "$failing_shell" | sed 's/-.*//')
+    failing_shell_version=$(echo "$failing_shell" | sed 's/.*-//')
+    if [ "$failing_shell_name" = "$shell" ]; then # First match on the shell name, then on the version if any
+      if [ -z "$failing_shell_version" ] || [ "$failing_shell_version" = "$failing_shell" ]; then
+        return 0 # No version specified, match!
+      elif shell_version "$shell" | grep -q -E "$failing_shell_version"; then
+        return 0 # version matched!
+      else
+        return 1 # version didn't match!
+      fi
+    fi
+  done
+  return 1 # No match
 }
 
-execute_test() { # executable: $1, args: $2 ...
+execute_test() { # executable: $1, timeout: $2, args: $3
   if [ "$backend" = "sh" ]; then
-    bash "./$1" $2 # Default to bash for sh backend
+    # Use a 30s timeout to prevent infinite loops
+    timeout ${2:-30} $shell "./$1" $3
```

Comment:
This script used to ignore the `--shell` option and would always use bash. I fixed the issue, but this meant I had to disable a failing few tests for `yash`.